File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, n: int, prerequisites: List[List[int]]) -> bool:
        
        # adj = defaultdict(list)
        # for u, v in prerequisites:
        #     adj[u].append(v)
        # print("ADj :", adj)

        # visited = [0] * n

        # def dfs(node):
        #     if visited[node] == 1:
        #         return False
        #     if visited[node] == 2:
        #         return True

        #     visited[node] = 1

        #     for neighbor in adj[node]:
        #         if not dfs(neighbor):
        #             return False

        #     visited[node] = 2
        #     return True

        # for course in range(n):
        #     if visited[course] == 0:
        #         if not dfs(course):
        #             return False

        # return True

        adj = defaultdict(list)
        indegree = [0] * n
        for course, prereq in prerequisites:
            indegree[course] += 1
            adj[prereq].append(course)

        # print("ADJ : ", adj)
        # print("IN : ", indegree)

        q = deque()
        for i in range(n):
            if indegree[i] == 0:
                q.append(i)
        
        topo = []

        while q:
            node = q.popleft()

            topo.append(node)
            for neighbor in adj[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    q.append(neighbor)

        # print("TOPO : ", topo)

        return len(topo) == n

File: 207-course-schedule/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        

        List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            adjList.add(new ArrayList<>());
        }

        int[] indegree = new int[numCourses];

        for (int[] prereq : prerequisites) {
            int course = prereq[0];
            int prerequisite = prereq[1];

            adjList.get(prerequisite).add(course);
            indegree[course]++;
        }

        Queue<Integer> q = new LinkedList<>();

        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) {
                q.offer(i);
            }
        }

        int nodes = 0;

        while(!q.isEmpty()) {
            int node = q.poll();
            nodes++;

            for (int n : adjList.get(node)) {
                indegree[n]--;
                if (indegree[n] == 0) {
                    q.offer(n);
                }
            }
        }

        System.out.println("Print, total nodes => " + nodes);
        return nodes == numCourses;
    }
}  